### ISSUE TITLE
Adding parameters to query

### ISSUE THREAD

**Issue by booni3** [matchish/laravel-scout-elasticsearch#86]
I see the package is using the ONGR `QueryStringQuery` by default and a lot is already possible with the [elasticsearch query syntax](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html#query-string-syntax).

In my case, I simply want to set a searchable fields array to the query but cannot see an easy way to append parameters to the ONGR query.

```json
{
    "query_string" : {
        "fields" : ["name", "email"],
        "query" : "this AND that OR thus"
    }
}
```

Is it possible to amend the query inside the scout search callback, or would this need amendment to the search factory to inject the additional paramaters here?

I know you can get the queries from within the callback with `$body->getQueries()` but when filters are being used I am not sure its possible to determine which query is from `$builder->query`... unless we assume there is only going to be a single `QueryStringQuery` but that does not feel safe.

**Comment by matchish** [comment]
Yes, it's an issue. Probably it's not a good idea to pass a body to a callback and maybe it's better to pass builder itself and build the body from scratch. But I'm not sure about the drawbacks of the solution now. 
https://github.com/matchish/laravel-scout-elasticsearch/blob/f2cf4d79fc8dfbe0e712dc57fd45a5bc43bd416e/src/Engines/ElasticSearchEngine.php#L141

**Comment by booni3** [comment]
I added an idea in #85 but it is a breaking change

**Comment by matchish** [comment]
Here the same solution as here https://github.com/matchish/laravel-scout-elasticsearch/issues/88#issuecomment-585557200
If we have callback in builder just send all builder and `elasticsearch` client to the callback then callback can build the right query from scratch
```
$model::search($term, function ($builder, $elasticsearch) {
  $query = new ElasticsearchMQuery::fromBuilder();
  return $elasticsearch->msearch($query->toArray());
})->get();
```
Feel free to send PR if you want.

Comments and objections are welcome)

**Comment by matchish** [comment]
This works today — no library change is needed.

`SearchFactory` passes any `BuilderInterface` value straight to the bool query instead of wrapping it in a term query, and `QueryStringQuery` takes a parameters array as its second constructor argument. So you can set `fields` like this:

```php
use ONGR\ElasticsearchDSL\Query\FullText\QueryStringQuery;

Product::search()
    ->where('q', new QueryStringQuery('this AND that', [
        'fields' => ['name', 'email'],
    ]));
```

That produces:

```json
{"query":{"bool":{"filter":[{"query_string":{"query":"this AND that","fields":["name","email"]}}]}}}
```

The `'q'` key is only a placeholder — it is ignored when the value is a DSL object.

One thing to know: `where()` adds the query in **filter** context, so Elasticsearch does not score the hits. Every result comes back with `_score: 0` and the order is not by relevance. If you need ranking, use the search callback, where `$body` is the ONGR `Search` object:

```php
Product::search('', function (\Elastic\Elasticsearch\Client $client, $body) {
    $body->addQuery(new QueryStringQuery('this AND that', [
        'fields' => ['name', 'email'],
    ]));

    return $client->search(['index' => 'products', 'body' => $body->toArray()])->asArray();
})->raw();
```

This puts the query at the top level, so scoring works normally.
